src/qibo/gates/abstract.py

```diff
@@ -359,18 +360,95 @@ def controlled_by(self, *qubits: int) -> "Gate":
             self.control_qubits = qubits
         return self
 
-    def decompose(self, *free) -> List["Gate"]:
+    def _base_decompose(self, *free, use_toffolis=True) -> List["Gate"]:
+        """Base decomposition for gates.
+
+        Returns a list containing the gate itself. Should be overridden by subclasses
+        that support decomposition to simpler gates.
+
+        Args:
+            free: Ids of free qubits to use for the gate decomposition.
+
+        Returns:
+            list: gates that have the same effect as applying the original gate.
+        """
+        return [self.__class__(*self.init_args, **self.init_kwargs)]
+
+    @staticmethod
+    def gates_cancel(g1, g2):
```

Comment:
```suggestion
    @staticmethod
    def _gates_cancel(g1, g2):
```
It seems that this method does not need to be user-facing.